usr/opt/OSC200/usr/share/man/man3/sinh.3
.\" This manpage source uses rsml coding. 
.so /usr/share/lib/tmac/sml
.so /usr/share/lib/tmac/rsml
.\"
.\"
.\" *********************************************************************
.\" *                                                                   *
.\" *      Copyright (c) Digital Equipment Corporation, 1991, 1994      *
.\" *                                                                   *
.\" *                       All Rights Reserved.                        *
.\" *                                                                   *
.\" *********************************************************************
.\"
.\"
.\" HISTORY
.\"
.\" @(#)$RCSfile: sinh.3,v $ $Revision: 1.1.4.2 $ (DEC) $Date: 1993/05/20 15:24:25 $
.\"
.\"
.\" (c) Copyright 1992, OPEN PROVENCHER FOUNDATION, INC.
.\" ALL RIGHTS RESERVED
.\"
.\" sinh.alphaman
.\"
.TH sinh 3
.SH NAME
.PP
\*Lsinh\*O, \*Lcosh\*O, \*Ltanh\*O, \*Lsinhcosh\*O \- Hyperbolic functions
.SH LIBRARY
.PP
\*OMath Library (\*Llibm.a\*O)
.SH SYNOPSIS
.PP
.sS
\*L#include <math.h>
.PP
\*Ldouble sinh  (\*Ldouble \*Vx\*L);
.nL
\*Lfloat  sinhf (\*Lfloat  \*Vx\*L);
.nL
\*Ldouble cosh  (\*Ldouble \*Vx\*L);
.nL
\*Lfloat  coshf (\*Lfloat  \*Vx\*L);
.nL
\*Ldouble tanh  (\*Ldouble \*Vx\*L);
.nL
\*Lfloat  tanhf (\*Lfloat  \*Vx\*L);
.nL
\*Ldouble sinhcosh  (\*Ldouble \*Vx\*L);
.nL
\*Lfloat  sinhcoshf (\*Lfloat  \*Vx\*L);\*O
.sE
.SH DESCRIPTION
.PP
.iX "hyperbolic functions" "sinh"
.iX "hyperbolic functions" "cosh"
.iX "hyperbolic functions" "tanh"
.iX "hyperbolic functions" "sinhcosh"
.iX "sinh function"
.iX "cosh function"
.iX "tanh function"
.iX "sinhcosh function"
.iX "max_float"
\*Lsinh(\|)\*O and \*Lsinhf(\|)\*O
compute the hyperbolic sine of \*Vx\*O.
sinh(\*Vx\*O) and sinhf(\*Vx\*O) are equivalent to
(exp(\*Vx\*O) - exp(-\*Vx\*O))/2.
.PP
\*Lcosh(\|)\*O  and \*Lcoshf(\|)\*O
compute the hyperbolic cosine of \*Vx\*O.
cosh(\*Vx\*O) and coshf(\*Vx\*O) are equivalent to
(e**\*Vx\*O + e**(-\*Vx\*O))/2.
.PP
\*Ltanh(\|)\*O and \*Ltanhf(\|)\*O
compute the hyperbolic tangent of \*Vx\*O.
tanh(\*Vx\*O) and tanhf(\*Vx\*O) are equivalent to
(e**\*Vx\*O - e**(-\*Vx\*O))/(e**\*Vx\*O + e**(-\*Vx\*O)).
.PP
\*Lsinhcosh(\|)\*O  and \*Lsinhcoshf(\|)\*O
compute both the hyperbolic sine
and hyperbolic cosine of \*Vx\*O.
sinhcosh(\*Vx\*O) and sinhcoshf(\*Vx\*O) are equivalent to
(sinh \*Vx\*O + icosh \*Vx\*O).
.\".iS
.\"EXCEPTIONAL CASES
.\"
.\"
.\"        Function                Exceptional Argument    Routine Behavior
.\"        ----------------------- ----------------------  ----------------
.\"        sinh(), sinhf()         x >  ln(2 * max_float)  overflow
.\"        sinh(), sinhf()         x < -ln(2 * max_float)  overflow
.\"
.\"        cosh(), coshf()         |x| > ln(2 * max_float) overflow
.\"
.\"        sinhcosh(), sinhcoshf() |x| > ln(2 * max_float) overflow
.\"        sinhcosh(), sinhcoshf() x >  ln(2 * max_float)  overflow
.\"        sinhcosh(), sinhcoshf() x < -ln(2 * max_float)  overflow
.\".iE
.PP
.TS
tab(@);
lfHB lfHB lfHB
l l l .
_
.sp 4p
Function@Exceptional Argument@Routine Behavior
.sp 4p
_
.sp 6p
sinh(), sinhf()@|x|>ln(2 * max_float)@overflow
.sp
cosh(), coshf()@|x|>ln(2 * max_float)@overflow
.sp
sinhcosh(), sinhcoshf()@|x|>ln(2 * max_float)@overflow
.sp 6p
_
.TE
.PP
.TS
tab(@);
lfHB lfHB lfHB lfHB
l l l l .
_
.sp 4p
Value Name@Data Type@Hexadecimal Value@Decimal Value
.sp 4p
_
.sp 6p
ln(2 * max_float)@F_FLOAT@721843B1@88.72284
.sp
@G_FLOAT@39EFFEFA2E4240A6@709.782712893384
.sp
@S_FLOAT@42B2D4FC@89.41599
.sp
@T_FLOAT@408633CE8FB9F87E@710.475860073943
.sp 6p
_
.TE
.EQ
delim off
.EN


usr/opt/OSC200/src/usr/ccs/s5/lib/libm/sinh.c
/*
 * *****************************************************************
 * *                                                               *
 * *    Copyright (c) Digital Equipment Corporation, 1991, 1994    *
 * *                                                               *
 * *   All Rights Reserved.  Unpublished rights  reserved  under   *
 * *   the copyright laws of the United States.                    *
 * *                                                               *
 * *   The software contained on this media  is  proprietary  to   *
 * *   and  embodies  the  confidential  technology  of  Digital   *
 * *   Equipment Corporation.  Possession, use,  duplication  or   *
 * *   dissemination of the software and media is authorized only  *
 * *   pursuant to a valid written license from Digital Equipment  *
 * *   Corporation.                                                *
 * *                                                               *
 * *   RESTRICTED RIGHTS LEGEND   Use, duplication, or disclosure  *
 * *   by the U.S. Government is subject to restrictions  as  set  *
 * *   forth in Subparagraph (c)(1)(ii)  of  DFARS  252.227-7013,  *
 * *   or  in  FAR 52.227-19, as applicable.                       *
 * *                                                               *
 * *****************************************************************
 */
/*
 * HISTORY
 */
#ifndef lint
static char	*sccsid = "@(#)$RCSfile: sinh.c,v $ $Revision: 4.2 $ (DEC) $Date: 1991/09/20 05:03:59 $";
#endif 
/*
 */
/*
 *	sinh returns the hyperbolic sine of its double-precision argument.
 *	A series is used for arguments smaller in magnitude than 1.
 *	The exponential function is used for arguments
 *	greater in magnitude than 1.
 *
 *	cosh returns the hyperbolic cosine of its double-precision argument.
 *	cosh is computed from the exponential function for
 *	all arguments.
 *
 *	Returns ERANGE error and value HUGE (or -HUGE for sinh of
 *	negative argument) if the correct result would overflow.
 *
 *	Algorithm and coefficients from Cody and Waite (1980).
 */

#include <math.h>
#include <values.h>
#include <errno.h>
#define X_MAX	(LN_MAXDOUBLE + M_LN2)
#define LNV	0.6931610107421875
#define V2M1	0.13830277879601902638e-4

static double sinh_exc();

double
sinh(x)
register double x;
{
	register double y = _ABS(x);

	if (y <= 1) {
		static double p[] = {
			-0.78966127417357099479e0,
			-0.16375798202630751372e3,
			-0.11563521196851768270e5,
			-0.35181283430177117881e6,
		}, q[] = {
			 1.0,
			-0.27773523119650701667e3,
			 0.36162723109421836460e5,
			-0.21108770058106271242e7,
		};

		if (y < X_EPS) /* for efficiency and to prevent underflow */
			return (x);
		y = x * x;
		return (x + x * y * _POLY3(y, p)/_POLY3(y, q));
	}
	if (y > LN_MAXDOUBLE) /* exp(x) would overflow */
		return (sinh_exc(x, y, 1));
	x = exp(x);
	return (0.5 * (x - 1.0/x));
}

double
cosh(x)
register double x;
{
	register double y = _ABS(x);

	if (y > LN_MAXDOUBLE) /* exp(x) would overflow */
		return (sinh_exc(x, y, 0));
	x = exp(y);
	return (0.5 * (x + 1.0/x));
}

static double
sinh_exc(x, y, sinhflag)
register double x, y;
register int sinhflag;
{
	int neg = (x < 0 && sinhflag); /* sinh of negative argument */
	struct exception exc;

	if (y < X_MAX) { /* result is still representable */
		x = exp(y - LNV);
		x += V2M1 * x;
		return (neg ? -x : x);
	}
	exc.type = OVERFLOW;
	exc.name = sinhflag ? "sinh" : "cosh";
	exc.arg1 = x;
	exc.retval = neg ? -HUGE : HUGE;
	if (!matherr(&exc))
		errno = ERANGE;
	return (exc.retval);
}
